File: typescript/src/giftcard/application/integration/booking/handlers/refundFromBooking.ts
import type { BookingRejectedIntegrationEvent } from '@/booking/integration/giftcard/bookingResultIntegrationEvent.js';
import { BookingResult } from '@/giftcard/application/integration/booking/adapter/bookingResult.js';
import { GiftCardRefunding } from '@/giftcard/application/usecases/giftCardRefunding.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class RefundFromBooking {
  private readonly _bookingResult: BookingResult;
  private readonly _useCase: GiftCardRefunding;

  constructor(bookingResult: BookingResult, giftCardRefunding: GiftCardRefunding) {
    requireDependency(bookingResult, 'bookingResult');
    this._bookingResult = bookingResult;
    requireDependency(giftCardRefunding, 'giftCardRefunding');
    this._useCase = giftCardRefunding;
  }

  handle(event: BookingRejectedIntegrationEvent): void {
    const command = this._bookingResult.adaptRejected(event);

    if (command !== null) {
      this._useCase.invoke(command);
    }
  }
}

File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/integration/booking/handlers/RefundFromBooking.java
package it.giannibombelli.wsc2026.giftcard.application.integration.booking.handlers;

import it.giannibombelli.wsc2026.booking.integration.giftcard.BookingResultIntegrationEvent;
import it.giannibombelli.wsc2026.giftcard.application.commands.RefundGiftCard;
import it.giannibombelli.wsc2026.giftcard.application.integration.booking.adapter.BookingResult;
import it.giannibombelli.wsc2026.giftcard.application.usecases.GiftCardRefunding;

import static java.util.Objects.requireNonNull;

public final class RefundFromBooking {
    private final BookingResult bookingResult;
    private final GiftCardRefunding useCase;

    public RefundFromBooking(BookingResult bookingResult, GiftCardRefunding giftCardRefunding) {
        this.bookingResult = requireNonNull(bookingResult);
        this.useCase = requireNonNull(giftCardRefunding);
    }

    public void handle(BookingResultIntegrationEvent.BookingRejectedIntegrationEvent event) {
        final RefundGiftCard cmd = bookingResult.adaptRejected(event);

        if (cmd != null) {
            useCase.invoke(cmd);
        }
    }
}
